### ISSUE TITLE
Always directing to wrong /login page when using behinde a proxy such as traefik

### ISSUE THREAD

**Issue by xalexslx** [reloxx13/TasmoAdmin#183]
I'm using traefik (similar to nginx with letsencrypt ans another features) to be the edge router of all my accessable containers on docker.

However, tasmoadmin is not accessible if my rule is to use tasmoadmin as a path of my domain, eg:
home.mydomain.com/tasmoadmin as it automatically redirects me to home.mydomain.com/login instead of home.mydomain.com/tasmoadmin/login.

**Comment by kevsgithub** [comment]
I'm not using traefik but surely it's just a case of editing .htaccess ?

I assume anyone not using the web root as a default would need to do this.

**Comment by xalexslx** [comment]
I think it's the case of anyone using any reverse proxy. I use traefik because it plays well with docker environments but the same is going to happen with nginx or any reverse proxy. Most applications works fine because they probably read the current path from the request and works from there.

**Comment by RaymondMouthaan** [comment]
Hai guys,

Internally the TasmoAdmin docker image uses nginx with the following config:

```
worker_processes  1;
pid /var/run/nginx.pid;
user nginx nginx;

events {
    worker_connections  1024;
}

http {
    include            mime.types;
    default_type       application/octet-stream;
    sendfile           on;
    keepalive_timeout  65;

    server {
        server_name tasmoadmin.local;
        listen 80 default_server;
        listen [::]:80 default_server;
        root /var/www/tasmoadmin/;
        index index.php;

        fastcgi_buffer_size 512k;
        fastcgi_buffers 16 512k;

        
        location /data/firmwares {
        }

        location /data/ {
            deny all;
        }

        location ~ .php$ {
            fastcgi_pass 127.0.0.1:9001;
            fastcgi_read_timeout 900;
            fastcgi_split_path_info ^(.+\.php)(/.+)$;
            fastcgi_index index.php;
            fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
            include fastcgi_params;
        }

        location ~ ^(.*)\.(css|js|gif||jpe?g|png|json|cache\.json)$ {
        }

        location / {
            rewrite ^/login$ /login.php last;
            rewrite ^/logout$ /login.php?logout=logout last;
            rewrite ^/doAjaxAll$ /index.php?doAjaxAll=doAjaxAll last;
            rewrite ^/doAjax$ /index.php?doAjax=doAjax last;
            rewrite "/([a-z]{2})/" /index.php?lang=$1 last;
            rewrite ^/([a-zA-Z_]+)/([a-zA-Z_]+)/([0-9_]+)/?$ /index.php?page=$1&action=$2&device_id=$3;
            rewrite ^/([a-zA-Z_]+)/(force)/?$ /index.php?page=$1&force=1;
            rewrite ^/([a-zA-Z_]+)/([a-zA-Z_]+)/?$ /index.php?page=$1&action=$2;
            rewrite ^/([a-zA-Z_]+)/([0-9]+)/?$ /index.php?page=$1&device_id=$2;
            rewrite ^/([a-zA-Z_]+)/?$ /index.php?page=$1;
        }
    }
}
```

When running on port 8091 my url becomes http://tasmoadmin.mydomain:8091, using haproxy as a reverse proxy this works very well.

Do I need to change / add anything to the nginx.config to make it work in the url format like http://my.domain/tasmoadmin?

Please advise :)

Cheers,
Ray

**Comment by kevsgithub** [comment]
No because the regex matches the start of the url. If you remove the '^/' on the regex then it would match on any non tasmoadmin stuff also on the server. If people want a custom install then they will need to edit the config to their requirements.

eg : 

`rewrite ^/tasmoadmin/([a-zA-Z_]+)/?$ /tasmoadmin/index.php?page=$1;`

the only way tasmoadmin could do anything is to force installation into a subdirectory for everybody.

**Comment by RaymondMouthaan** [comment]
What I also could do is making it possible for users to supply their own custom nginx config. That would make it possible to configure nginx fully. This file could be stored outside the container and after a re-deployment applied again. Currently this nginx conf is stored inside the container, when a user customizes it, all changes will be lost after a new deployment of the container.

**Comment by reloxx13** [comment]
im already getting the subdir and also im using it .


im getting it form  $_SERVER[ 'PHP_SELF' ]

i would need an output of your $_SERVER variable, cuzi dont have a proxy.

add this in your includes/top.php before the $subdir block, like this:



```
	debug( $_SERVER );
	die();
	$subdir = dirname( $_SERVER[ 'PHP_SELF' ] )."/";
	$subdir = $subdir = str_replace( "\\", "/", $subdir );
	$subdir = $subdir == "//" ? "/" : $subdir;
```

**Comment by RaymondMouthaan** [comment]
@reloxx13, but that would mean users needed to modify TasmoAdmin inside the container, which I prefer to avoid... Right?

**Comment by reloxx13** [comment]
yep, but this wont solve the issue since its stil not physically in a sub dir.

from my researches i found examples like this:
`proxy_pass http://localhost:3200/;`

this could fix it, but i cannot test without a proxy 😭

**Comment by RaymondMouthaan** [comment]
You can, just use docker to deploy a proxy 😌

**Comment by reloxx13** [comment]
haha dude im on it already 🤣 🚀 

https://www.digitalocean.com/community/tutorials/how-to-use-traefik-as-a-reverse-proxy-for-docker-containers-on-ubuntu-16-04

**Comment by ConfusedTA** [comment]
I'm trying to set it up on an Apache reverse proxy, so it'll be https://url/tasmoadmin but rather than redirecting to https://url/tasmoadmin/login it's trying to redirect to https://url/login

Other apps I've got set up this way have a base_url property that can be set by the user to allow reverse proxying.

**Comment by tux1337** [comment]
I would love to see the proposed base_url property from @ConfusedTA  like other application does this. 

Normally things like that are configurable over an environment variable in the docker world. Linking an custom nginx.conf into the container is a little bit to take a sledgehammer to crack a nut.

**Comment by caius88** [comment]
Hi @reloxx13 nice project! 
I have the same issue described above tring to proxy TasmoAdmin with Traefik :(
Is there any hope do you fix it in the short time? Or any workaround/config?

Thanks in advance :)

**Comment by darnokb** [comment]
Any news on this? I am also struggling with the same issue. Any workaroung/fix ready?

**Comment by cmosek** [comment]
Please provide `base_url` property that can be set by the user to allow reverse proxying.

**Comment by inverse** [comment]
Is this still an issue? there's a `TASMO_BASEURL` environment variable that can be set to control this from what I understand.

**Comment by reloxx13** [comment]
I checked current master before changing anything: TASMO_BASEURL is read in includes/bootstrap.php, login/logout redirects go through _BASEURL_, and I verified locally that _BASEURL_ becomes /tasmoadmin/ when that env var is set. I could not reproduce an active code defect from the current tree alone.

If this still happens on a current image, please share:
- the exact TasmoAdmin version or image tag
- whether TASMO_BASEURL is set, and to what value
- the reverse-proxy path rule / rewrite config
- the exact URL requested and the exact redirect target you receive

With that, it should be possible to tell whether there is still a live bug here or a deployment/config mismatch.

**Comment by reloxx13** [comment]
Triage note: I checked the current codebase before closing this. `tasmoadmin/includes/bootstrap.php` now supports `TASMO_BASEURL`, and login/logout redirects consistently go through `_BASEURL_`, so the original blanket path-prefix redirect problem is no longer evident from the current implementation. This old investigate thread is being closed unless someone can reproduce it on a current image with exact proxy configuration details.
